modules/mono/glue/GodotSharp/GodotSharp/Core/Array.cs

```diff
@@ -1137,6 +1156,16 @@ public Array(Array array)
                 throw new ArgumentNullException(nameof(array));
 
             _underlyingArray = array;
```

Comment:
I personally think that the constructor _should_ shallow copy, because right now you can have the odd situation where you have two collections that are not the same according to `object.ReferenceEquals`, but nonetheless refer to the same engine-side array and as such will observe changes made by each other.

However, changing that would be a breaking change with user-observable effects (IIRC this is relevant when there are multiple instances of the same scene) and at the very least I think it's not within scope for this PR.

Reply:
> it won't be able to fix cases where we have a existing saved untyped array as in https://github.com/godotengine/godot/issues/98309.

Now that I'm reading this again and recalling how exported collections are loaded, I think this PR might partially fix #98309.

https://github.com/godotengine/godot/blob/f8b5698ecfe6060a14077021fa99319883f8126c/scene/resources/packed_scene.cpp#L565-L570
Here, what the scene loader does is duplicate the value that the exported field already has, which copies over its element type if it's set, and then sets the elements to those that are stored in the scene.

I think there still is a bug with this when the default value of the field is null. I thought that this was only an issue with exported dictionaries, but seeing this I might be wrong about that. See #98546

Reply:
Yeah I agree it should be the behavior but given the current use it would be both a breaking change and make existing code paths of native -> c# array slower; maybe godot 5 lol? I think the current constructors with that behavior could be made internal and explicitly for cases like `CreateTakingOwnershipOfDisposableValue`, and then the public constructor making a shallow copy still benefits over the c# collection constructors since it knows it can make a single api call to duplicate rather than 2N calls to read/insert.